### DISCUSSION TITLE
Using SWR with use effect causes Error: Rendered more hooks than during the previous render.

### DISCUSSION THREAD

**Discussion by adherb** [vercel/swr#1815] [Q&A] [answered]
I am using SWR to get data from an api, I am then passing the setState (setTotalRecipes) from another page to this page and I would like to setState for data returned from the SWR fetch, but my useEffect keeps triggering this error:  Error: Rendered more hooks than during the previous render.

Here is my code:
```
export default function RecipeList({ setTotalRecipes }) {

 const { data, error } = useSWR(
    `${process.env.NEXT_PUBLIC_SERVER_API}/recipes/get?skip=${toSkip}&take=16`,
    fetcher
  );

 if (error) return <div>failed to load</div>;
 if (!data) return <divLoading...</div>;

useEffect(() => {
    if (data.totalRecipes) {
      setTotalRecipes(data.totalRecipes);
    }
  }, [data.totalRecipes]);

return();
}
```

How can I get this to work?

**Comment by icyJoseph** [comment, marked as answer]
Hi, 

This is done by react. The number of hooks in the body of a functional component cannot change between renders.

If we statically analyze your code:

```tsx
export default function RecipeList({ setTotalRecipes }) {

 const { data, error } = useSWR(
    `${process.env.NEXT_PUBLIC_SERVER_API}/recipes/get?skip=${toSkip}&take=16`,
    fetcher
  );

 if (error) return <div>failed to load</div>;
 if (!data) return <divLoading...</div>;

useEffect(() => {
    if (data.totalRecipes) {
      setTotalRecipes(data.totalRecipes);
    }
  }, [data.totalRecipes]);

return();
}
```

We see that if `error` is truthy, then the `useEffect` is not called, so it is not seen by React. Because this `if` branch ends the function execution.

The same applies if `error` is falsy, but `!data` is truthy.

You need to move your `useEffect` upwards, and potentially add one more check to the `useEffect` callback.


```jsx
export default function RecipeList({ setTotalRecipes }) {

 const { data, error } = useSWR(
    `${process.env.NEXT_PUBLIC_SERVER_API}/recipes/get?skip=${toSkip}&take=16`,
    fetcher
  );

useEffect(() => {
    if(!data) return;
    if (data.totalRecipes) {
      setTotalRecipes(data.totalRecipes);
    }
  }, [data]);

 if (error) return <div>failed to load</div>;
 if (!data) return <divLoading...</div>;

return();
}
```

**Comment by adherb** [comment]
Hi @icyJoseph,

Thank you for the reply, it worked perfectly. I really appreciate you taking the time to explain that to me.
